### DISCUSSION TITLE
Is it possible to render plain html content in head?

### DISCUSSION THREAD

**Discussion by migueldamota** [vercel/next.js#71032] [Help]
### Summary

We start transitioning from WordPress but we still need to load some custom css and script which aren't just the inner code but the full tag.
So is it possible to render these elements in the head server side?

I mean plain html content like this:
```html
<script>
console.log("hullu");
</script>

<!-- or -->

<style>
.a {
/* ... */
}
</style>
```

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.1.0: Tue Sep 17 07:46:59 PDT 2024; root:xnu-11215.40.59~38/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 8
Binaries:
  Node: 20.16.0
  npm: 10.8.1
  Yarn: 1.22.22
  pnpm: 8.15.9
Relevant Packages:
  next: 15.0.0-canary.181 // Latest available version is detected (15.0.0-canary.181).
  eslint-config-next: N/A
  react: 19.0.0-rc-2d16326d-20240930
  react-dom: 19.0.0-rc-2d16326d-20240930
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

**Comment by lumirlumir** [comment]
Do you use App Router or Pages Router?

Pages Router has `<Head>` component imported from `next/head`. I think it can be used in your case.

This method is good in Pages Router, But I am not sure this also works in App Router. Some optimizations can be missed if you use this.

- https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#step-3-migrating-nexthead

I think you can use it like the code below.

```javascript
import Head from 'next/head';

const MyPage = () => {
  return (
    <>
      <Head>
        <link rel="stylesheet" href="https://example.com/styles.css" />
        <script src="https://example.com/script.js" async></script>
      </Head>
      <div>
        {/* Contents */}
      </div>
    </>
  );
};

export default MyPage;
```

**Reply by migueldamota** [reply]
Hi @lumirlumir, I am using App Router. I may have explained it wrong, we have a plugin in WordPress which we use to add custom scripts, styles or full html content to either head or body (footer) and now I have an API endpoint where I get those scripts/styles/etc., but the plugin only returns the full html content.

The best would be something like this:

```jsx

export default async function Layout({ children }) {
    const scripts = await getScripts(); // returns plain html content (all scripts, styles, etc.)

    return (
        <html lang="en">
            <head>
                {scripts}
            </head>
            <body className="font-sans antialiased">{children}</body>
        </html>
    );
}
```

I can build a custom parser which parses each item on their html content and gets their content, links, href/src and add them with `<Script />` or `<style />` but this could be a bit complicated. If the approach mentioned above is not really working, I'll have to accept it and build a custom parser for that. Anyways thanks in advance! 🙏

**Reply by lumirlumir** [reply]
@migueldamota I tried to find another way, but I also think using a parser is the second-best option.

If my suggestion doesn’t work out, take a look at this JavaScript package [`cheerio`](https://github.com/cheeriojs/cheerio). With this package, you won’t need a custom parser. As far as I know, it can parse `<head>` tags and handles HTML strings in a jQuery-like way.

Hope you’re doing well :)

**Reply by migueldamota** [reply]
@lumirlumir I now have a custom parser in PHP, it is easier to use a html document parser which can then reads all attributes and content like jQuery 😅
Thanks for your time and answers 🙌
